File: scripts/check_markdown_links.py
#!/usr/bin/env python3
"""Check relative links in tracked Markdown files without external packages."""

from __future__ import annotations

import re
import subprocess
import sys
from pathlib import Path
from urllib.parse import unquote, urlsplit


INLINE_LINK_RE = re.compile(r"!?\[[^\]]*\]\(([^)]+)\)")
REFERENCE_LINK_RE = re.compile(r"^\s{0,3}\[[^\]]+\]:\s*(.+?)\s*$")
FENCE_RE = re.compile(r"^\s*(`{3,}|~{3,})")


def tracked_markdown_files(repo_root: Path) -> list[Path]:
    result = subprocess.run(
        ["git", "ls-files", "-z", "--", "*.md"],
        cwd=repo_root,
        check=True,
        stdout=subprocess.PIPE,
    )
    return [repo_root / path.decode() for path in result.stdout.split(b"\0") if path]


def link_destination(raw: str) -> str:
    raw = raw.strip()
    if raw.startswith("<"):
        end = raw.find(">")
        return raw[1:end] if end != -1 else raw
    return raw.split(maxsplit=1)[0] if raw else ""


def is_relative_file_link(destination: str) -> bool:
    if not destination or destination.startswith(("#", "/")):
        return False

    parsed = urlsplit(destination)
    return not parsed.scheme and not parsed.netloc and bool(parsed.path)


def check_file(repo_root: Path, markdown_file: Path) -> list[str]:
    errors: list[str] = []
    in_fence = False
    fence_char = ""
    fence_len = 0

    for line_number, line in enumerate(markdown_file.read_text(encoding="utf-8").splitlines(), 1):
        fence_match = FENCE_RE.match(line)
        if fence_match:
            marker = fence_match.group(1)
            if not in_fence:
                in_fence = True
                fence_char = marker[0]
                fence_len = len(marker)
            elif marker[0] == fence_char and len(marker) >= fence_len:
                in_fence = False
            continue

        if in_fence:
            continue

        destinations = [match.group(1) for match in INLINE_LINK_RE.finditer(line)]
        reference_match = REFERENCE_LINK_RE.match(line)
        if reference_match:
            destinations.append(reference_match.group(1))

        for raw_destination in destinations:
            destination = link_destination(raw_destination)
            if not is_relative_file_link(destination):
                continue

            path = unquote(urlsplit(destination).path)
            target = (markdown_file.parent / path).resolve()
            try:
                target.relative_to(repo_root)
            except ValueError:
                errors.append(
                    f"{markdown_file.relative_to(repo_root)}:{line_number}: "
                    f"relative link escapes repository: {destination}"
                )
                continue

            if not target.exists():
                errors.append(
                    f"{markdown_file.relative_to(repo_root)}:{line_number}: "
                    f"relative link target not found: {destination}"
                )

    return errors


def main() -> int:
    repo_root = Path(__file__).resolve().parent.parent
    errors: list[str] = []

    for markdown_file in tracked_markdown_files(repo_root):
        if markdown_file.is_file():
            errors.extend(check_file(repo_root, markdown_file))

    if errors:
        print("Broken Markdown relative links:", file=sys.stderr)
        for error in errors:
            print(f"  {error}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check_markdown_links.exs
#!/usr/bin/env elixir

defmodule MarkdownLinkCheck do
  @moduledoc false

  @inline_link_re ~r/!?\[[^\]]*\]\(([^)]+)\)/
  @reference_link_re ~r/^\s{0,3}\[[^\]]+\]:\s*(.+?)\s*$/
  @fence_re ~r/^\s*(`{3,}|~{3,})/

  def run do
    repo_root = Path.expand("..", __DIR__)

    errors =
      repo_root
      |> tracked_markdown_files()
      |> Enum.filter(&File.regular?/1)
      |> Enum.flat_map(&check_file(repo_root, &1))

    case errors do
      [] ->
        0

      errors ->
        IO.puts(:stderr, "Broken Markdown relative links:")
        Enum.each(errors, &IO.puts(:stderr, "  #{&1}"))
        1
    end
  end

  defp tracked_markdown_files(repo_root) do
    {output, 0} = System.cmd("git", ["ls-files", "-z", "--", "*.md"], cd: repo_root)

    output
    |> String.split(<<0>>, trim: true)
    |> Enum.map(&Path.join(repo_root, &1))
  end

  defp check_file(repo_root, markdown_file) do
    markdown_file
    |> File.stream!()
    |> Stream.with_index(1)
    |> Enum.reduce({[], nil}, fn {raw_line, line_number}, {errors, fence} ->
      line = trim_line_ending(raw_line)

      case Regex.run(@fence_re, line, capture: :all_but_first) do
        [marker] ->
          {errors, update_fence(fence, marker)}

        nil when not is_nil(fence) ->
          {errors, fence}

        nil ->
          line_errors = check_line(repo_root, markdown_file, line_number, line)
          {Enum.reverse(line_errors, errors), nil}
      end
    end)
    |> elem(0)
    |> Enum.reverse()
  end

  defp trim_line_ending(line) do
    line
    |> String.trim_trailing("\n")
    |> String.trim_trailing("\r")
  end

  defp update_fence(nil, marker), do: {String.first(marker), String.length(marker)}

  defp update_fence({fence_char, fence_len} = fence, marker) do
    if String.first(marker) == fence_char and String.length(marker) >= fence_len do
      nil
    else
      fence
    end
  end

  defp check_line(repo_root, markdown_file, line_number, line) do
    inline_destinations =
      @inline_link_re
      |> Regex.scan(line, capture: :all_but_first)
      |> Enum.map(&List.first/1)

    reference_destinations =
      case Regex.run(@reference_link_re, line, capture: :all_but_first) do
        [destination] -> [destination]
        nil -> []
      end

    (inline_destinations ++ reference_destinations)
    |> Enum.map(&link_destination/1)
    |> Enum.filter(&relative_file_link?/1)
    |> Enum.flat_map(fn destination ->
      check_destination(repo_root, markdown_file, line_number, destination)
    end)
  end

  defp link_destination(raw) do
    raw = String.trim(raw)

    cond do
      raw == "" ->
        ""

      String.starts_with?(raw, "<") ->
        case :binary.match(raw, ">") do
          {index, 1} -> binary_part(raw, 1, index - 1)
          :nomatch -> raw
        end

      true ->
        case String.split(raw, ~r/\s+/, parts: 2, trim: true) do
          [destination | _] -> destination
          [] -> ""
        end
    end
  end

  defp relative_file_link?(destination) do
    if destination == "" or String.starts_with?(destination, ["#", "/"]) do
      false
    else
      uri = URI.parse(destination)
      is_nil(uri.scheme) and is_nil(uri.host) and uri.path not in [nil, ""]
    end
  end

  defp check_destination(repo_root, markdown_file, line_number, destination) do
    path = destination |> URI.parse() |> Map.fetch!(:path) |> URI.decode()
    target = Path.expand(path, Path.dirname(markdown_file))
    source = Path.relative_to(markdown_file, repo_root)

    cond do
      not inside_repo?(target, repo_root) ->
        ["#{source}:#{line_number}: relative link escapes repository: #{destination}"]

      not File.exists?(target) ->
        ["#{source}:#{line_number}: relative link target not found: #{destination}"]

      true ->
        []
    end
  end

  defp inside_repo?(target, repo_root) do
    target == repo_root or String.starts_with?(target, repo_root <> "/")
  end
end

System.halt(MarkdownLinkCheck.run())
